### PR TITLE
optimize note performance

### PR DESCRIPTION
**Pull request by Gary-Hobson**
## Summary
Use sched_note_dump to record the original data to reduce calling printf during runtime

### CODE REVIEW

File: drivers/note/noteram_driver.c [hunk 1]
```diff
@@ -986,14 +986,48 @@ static int noteram_dump_one(FAR uint8_t *p, FAR struct lib_outstream_s *s,
...
+        switch (nbi->nbi_event)
           {
-            ret += lib_sprintf(s, " 0x%x", nbi->nbi_data[i]);
-          }
-
-        ret += lib_sprintf(s, "\n");
+            case NOTE_BEGINEX_EVENT:
+              {
+                ret += lib_sprintf(s, "tracing_mark_write: B|%d|%*s\n", pid,
+                                  count, nbi->nbi_data);
```
Comment (acassis) [inline drivers/note/noteram_driver.c:994]:
```suggestion
                                   count, nbi->nbi_data);

File: include/nuttx/sched_note.h [hunk 2]
```diff
@@ -136,14 +141,24 @@
           sched_note_bprintf_ip(tag, SCHED_NOTE_IP, event, \
                                 fmt, ##__VA_ARGS__)
 #  define sched_note_beginex(tag, str) \
-          sched_note_printf_ip(tag, SCHED_NOTE_IP, "B|%d|%s", _SCHED_GETTID(), str)
+          sched_note_dump_ip(tag, SCHED_NOTE_IP, NOTE_BEGINEX_EVENT, \
```
Comment (xiaoxiang781216) [inline include/nuttx/sched_note.h:144]: let's call sched_note_dump instead sched_note_dump_ip

File: include/nuttx/sched_note.h [hunk 3]
```diff
@@ -119,6 +119,11 @@
   memset((s), 0, sizeof(struct note_filter_tag_s));
 #endif
 
+#define NOTE_BEGINEX_EVENT 0
```
Comment (xiaoxiang781216) [inline include/nuttx/sched_note.h:122]:
```
#define NOTE_EVENT_BEGIN   0
#define NOTE_EVENT_END     1
#define NOTE_EVENT_MARK    2
#define NOTE_EVENT_COUNTER 3
```

File: include/nuttx/sched_note.h [hunk 4]
```diff
@@ -136,14 +141,24 @@
...
+          sched_note_dump_ip(tag, SCHED_NOTE_IP, NOTE_MARK_EVENT, \
+                             str, strlen(str))
 #  define sched_note_counter(tag, name, value) \
-          sched_note_printf_ip(tag, SCHED_NOTE_IP, "C|%d|%s|%" PRId32, \
-                               _SCHED_GETTID(), name, value)
+          do \
+            { \
+              struct note_counter_s _counter_ = \
+                { \
+                  name, value \
```
Comment (xiaoxiang781216) [inline include/nuttx/sched_note.h:157]: name may disappear in noteram_dump_one, let's change to:
```
#define sched_note_counter(tag, name, value) \
             sched_note_bprintf(tag, NOTE_EVENT_COUNTER, "%lx%s", (unsigned long)value, name)
```
and implement %s inside sched_note_vbprintf.

File: include/nuttx/sched_note.h [hunk 5]
```diff
@@ -700,8 +636,7 @@ void sched_note_filter_tag(FAR struct note_filter_tag_s *oldf,
 }
 #endif
 
-#else /* CONFIG_SCHED_INSTRUMENTATION */
-
+#ifndef CONFIG_SCHED_INSTRUMENTATION
```
Comment (xiaoxiang781216) [inline include/nuttx/sched_note.h:664]: why replace #else with #ifndef

File: drivers/note/noteram_driver.c [hunk 6]
```diff
@@ -973,6 +973,39 @@ static int noteram_dump_one(FAR uint8_t *p, FAR struct lib_outstream_s *s,
       }
       break;
 
+    case NOTE_DUMP_BEGIN:
+    case NOTE_DUMP_END:
+    case NOTE_DUMP_BEGINEX:
+    case NOTE_DUMP_ENDEX:
+    case NOTE_DUMP_MARK:
+      {
+        FAR struct note_binary_s *nbi = (FAR struct note_binary_s *)p;
```
Comment (xiaoxiang781216) [inline drivers/note/noteram_driver.c:982]: remove nbi_event from note_binary_s

File: drivers/note/noteram_driver.c [hunk 7]
```diff
@@ -972,7 +972,44 @@ static int noteram_dump_one(FAR uint8_t *p, FAR struct lib_outstream_s *s,
...
+    case NOTE_DUMP_END:
+      {
+        FAR struct note_binary_s *nbi = (FAR struct note_binary_s *)p;
+        ret += noteram_dump_header(s, &nbi->nbi_cmn, ctx);
+        char c = note->nc_type == NOTE_DUMP_BEGIN ? 'B' : 'E';
+        int len = note->nc_length - sizeof(struct note_binary_s);
+        if (len > 0)
+          {
+            ret += lib_sprintf(s, "tracing_mark_write: %c|%d|%.*s\n",
+                               c, pid, len, (FAR char *)nbi->nbi_data);
```
Comment (xiaoxiang781216) [inline drivers/note/noteram_driver.c:985]: (FAR const char *)

File: drivers/note/noteram_driver.c [hunk 8]
```diff
@@ -972,7 +972,44 @@ static int noteram_dump_one(FAR uint8_t *p, FAR struct lib_outstream_s *s,
...
+          }
+      }
+      break;
+    case NOTE_DUMP_MARK:
+      {
+        int len = note->nc_length - sizeof(struct note_binary_s);
+        FAR struct note_binary_s *nbi = (FAR struct note_binary_s *)p;
+        ret += noteram_dump_header(s, &nbi->nbi_cmn, ctx);
+        ret += lib_sprintf(s, "tracing_mark_write: I|%d|%.*s\n",
+                           pid, len, (FAR char *)nbi->nbi_data);
```
Comment (xiaoxiang781216) [inline drivers/note/noteram_driver.c:1000]: (FAR const char *)

File: include/nuttx/sched_note.h (hunk 5 above)
Comment (Gary-Hobson) [reply include/nuttx/sched_note.h:674]: This is to simplify the header file. It uses too many macros, and many of them are unnecessary.

File: drivers/note/noteram_driver.c (hunk 6 above)
Comment (Gary-Hobson) [reply drivers/note/noteram_driver.c:979]: done

File: drivers/note/noteram_driver.c (hunk 7 above)
Comment (Gary-Hobson) [reply drivers/note/noteram_driver.c:985]: done

File: include/nuttx/sched_note.h [hunk 9]
```diff
@@ -64,100 +62,94 @@
...
+#define sched_note_dump(tag, event, buf, len) \
+        sched_note_dump_ip(tag, SCHED_NOTE_IP, event, buf, len)
+#define sched_note_vprintf(tag, fmt, va) \
+        sched_note_vprintf_ip(tag, SCHED_NOTE_IP, fmt, va)
+#define sched_note_vbprintf(tag, event, fmt, va) \
+        sched_note_vbprintf_ip(tag, SCHED_NOTE_IP, event, fmt, va)
+#define sched_note_printf(tag, fmt, ...) \
+        sched_note_printf_ip(tag, SCHED_NOTE_IP, fmt, ##__VA_ARGS__)
+#define sched_note_bprintf(tag, event, fmt, ...) \
           sched_note_bprintf_ip(tag, SCHED_NOTE_IP, event, \
```
Comment (xiaoxiang781216) [inline include/nuttx/sched_note.h]: align

**Review by jerpelea — APPROVED** [review]
please add commit message

File: drivers/note/noteram_driver.c [hunk 10]
```diff
@@ -972,7 +972,44 @@ static int noteram_dump_one(FAR uint8_t *p, FAR struct lib_outstream_s *s,
...
+        int len = note->nc_length - sizeof(struct note_binary_s);
+        if (len > 0)
+          {
+            ret += lib_sprintf(s, "tracing_mark_write: %c|%d|%.*s\n",
+                               c, pid, len, (FAR const char *)nbi->nbi_data);
+          }
+        else
+          {
+            ret += lib_sprintf(s, "tracing_mark_write: %c|%d|%pS\n",
+                                c, pid, (FAR void *)nbi->nbi_ip);
```
Comment (xiaoxiang781216) [inline drivers/note/noteram_driver.c:990]: align

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**